### DISCUSSION TITLE
Http browser client construction?

### DISCUSSION THREAD

**Discussion by Quafadas** [disneystreaming/smithy4s#253] [Q&A] [answered]
I've successfully instantiated an http server, swagger docs etc.  Which is cool. I figured I'd push my luck all the way to the end, and see if I could generate a typesafe http client in the browser in scala JS, too. 

```
  val helloClient: org.http4s.client.Client[IO] = FetchClientBuilder[IO].create
  val myClient: Resource[cats.effect.IO, HelloWorldService[cats.effect.IO]] = example.MyClient.helloWorldClient(helloClient)
```
To my astonishment, compiled in a scala JS project!

It fell over at the `linking ` stage though. 

This may well not be inside the intent / design of smithy4s - please feel free to say that, but I just want to check that I'm not missing an obvious trick / import / intent. 

```
Referring to non-existent class smithy4s.http4s.package$
  called from example.MyClient$.helloWorldClient(org.http4s.client.Client)cats.effect.kernel.Resource
  called from constructor frontend.Main$.<init>()void
  exported to JavaScript with @JSExport
involving instantiated classes:
  example.MyClient$
    instantiated from constructor frontend.Main$.<init>()void
      (already seen, not repeating call stack)
Referring to non-existent class smithy4s.http4s.package$ServiceOps
  called from example.MyClient$.helloWorldClient(org.http4s.client.Client)cats.effect.kernel.Resource
  called from constructor frontend.Main$.<init>()void
  exported to JavaScript with @JSExport
involving instantiated classes:
  example.MyClient$
    instantiated from constructor frontend.Main$.<init>()void
      (already seen, not repeating call stack)
Referring to non-existent method smithy4s.http4s.package$ServiceOps.simpleRestJson()smithy4s.http4s.SimpleProtocolBuilder$ServiceBuilder
```

Is this something which has been successfully done before? It is "intended" to be possible? 

# Pipe Dream
Instantiate the client with laminars `EventStream` instead of IO.

https://github.com/Quafadas/mill-full-stack

**Comment by Baccata** [comment]
Summoning @keynmol who in all likelihood has the most extensive experience with using smithy4s in a front-end environment

**Comment by keynmol** [comment, marked as answer]
Hehe, my time has come. From quick look at the build, it seems you have fallen into a common trap - in fact so common, I created a site for it: https://youforgotapercentagesignoracolon.com/#mill

Specifically here: https://github.com/Quafadas/mill-full-stack/blob/master/build.sc#L31-L33 you're adding JVM versions (note the single `:` instead of `::`) of Smithy4s dependencies to your frontend module.

To allay your doubts, it is indeed possible, as I'm writing an app with smithy4s on the frontend and it's already running somewhere on the internet :) just not ready for a release yet.

**Reply by Quafadas** [reply]
Oh no, I can't believe I got so close, and stuck out on that. 

I love the website! 

It is mind-blowing, that this works. Thanks.

**Reply by Quafadas** [reply]
> To allay your doubts, it is indeed possible, as I'm writing an app with smithy4s on the frontend and it's already running somewhere on the internet :) just not ready for a release yet.

If you do I shall be studying it with great interest!

**Reply by kubukoz** [reply]
when I forget the triple colon:

![image](https://user-images.githubusercontent.com/894884/171906005-8cdbb0da-149e-4c63-a7e4-e8c8c17a64dc.png)
